nucleof072rb/Core/Src/main.c

```diff
@@ -1,181 +1,225 @@
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
+  /* USER CODE BEGIN 2 */
+  //set the max size of the bit transmission
+  const uint16_t max_bit_size = 1023;
+  //set the timer period
+  const uint16_t timer_period = 6400;
+
+  //transmit buffer with start bit (0000 0001), bit flag (1000 0000), zero bit (0000 0000).
+    uint8_t tx_buffer[3] = {0x01, 0x80, 0x00};
+
+  //empty buffer to receive output
+  uint8_t rx_buffer[3] = {0};
+
+  uint8_t last_two_bits = 0;
+  uint16_t adc_out_val = 0;
+  uint16_t adc_count_val = 0;
+
+  //set the max and minimum periods
+  uint16_t max_period = timer_period*0.10;
+  uint16_t min_period = timer_period*0.05;
+
+  //Starts the timer
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1);
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+	  //sets Chip Select (CS) to low because it idles in high
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_3, GPIO_PIN_RESET);
+
+	  //send tx_buffer and receive rx_buffer
+	  HAL_SPI_TransmitReceive(&hspi1, tx_buffer, rx_buffer, 3, HAL_MAX_DELAY);
+
+	  //sets Chip Select (CS) to high
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_3, GPIO_PIN_SET);
+
+	  //mask all but the last two bits of the second eight bits
+	  last_two_bits = rx_buffer[1] & 0x03;
+	  //moves all last two bits over and adds the third eight bits to it
+	  adc_out_val = (last_two_bits << 8 | rx_buffer[2]);
+
+	  //ADC value moved into 5-10% range of clock period
+	  adc_count_val = (min_period + (max_period - min_period)*adc_out_val)/max_bit_size;

```

Comment:
optimize the formula: `( 0.05 + ((float)adc_input / max_val)*0.05 ) * __HAL_TIM_GET_AUTORELOAD(. . .)`